warehouse/oidc/models.py

```diff
@@ -0,0 +1,185 @@
+    # indicating any custom claims that are known to be present but are
+    # not checked as part of verifying the JWT.
+    __unchecked_claims__: Set[str] = set()
+
+    @classmethod
+    def all_known_claims(cls):
+        """
+        Returns all claims "known" to this provider.
+        """
+        return (
+            cls.__verifiable_claims__.keys()
+            | cls.__preverified_claims__
+            | cls.__unchecked_claims__
+        )
+
+    def verify_claims(self, signed_claims):
+        """
+        Given a JWT that has been successfully decoded (checked for a valid
+        signature and basic claims), verify it against the more specific
+        claims of this provider.
+        """
+
+        # Defensive programming: treat the absence of any claims to verify
+        # as a failure rather than trivially valid.
+        if not self.__verifiable_claims__:
+            return False
+
+        # All claims should be accounted for.
+        # The presence of an unaccounted claim is not an error, only a warning
+        # that the JWT payload has changed.
+        unaccounted_claims = signed_claims.keys() - self.all_known_claims()
+        if unaccounted_claims:
+            sentry_sdk.capture_message(
+                f"JWT for {self.__class__.__name__} has unaccounted claims: "
+                f"{unaccounted_claims}"
+            )
+
+        # Finally, perform the actual claim verification.
+        for claim_name, check in self.__verifiable_claims__.items():
+            # All verifiable claims are mandatory. The absence of a missing
+            # claim *is* an error, since it indicates a breaking change in the
+            # JWT's payload.
+            signed_claim = signed_claims.get(claim_name)
+            if signed_claim is None:
+                sentry_sdk.capture_message(
+                    f"JWT for {self.__class__.__name__} is missing claim: {claim_name}"
+                )
+                return False
+
+            if not check(getattr(self, claim_name), signed_claim):
+                return False
+
+        return True
+
+    @property
+    def provider_name(self):  # pragma: no cover
+        # Only concrete subclasses of OIDCProvider are constructed.
+        return NotImplemented
+
+
+class GitHubProvider(OIDCProvider):
+    __tablename__ = "github_oidc_providers"
+    __mapper_args__ = {"polymorphic_identity": "github_oidc_providers"}
+    __table_args__ = (
+        UniqueConstraint(
+            "repository_name",
+            "owner",
+            "workflow_name",
+            name="_github_oidc_provider_uc",
+        ),
+    )
+
+    id = Column(UUID(as_uuid=True), ForeignKey(OIDCProvider.id), primary_key=True)
+    repository_name = Column(String)
+    owner = Column(String)
+    owner_id = Column(String)
+    workflow_name = Column(String)
+
+    __verifiable_claims__ = {
+        "repository": str.__eq__,
+        "job_workflow_ref": str.startswith,
```

Comment:
This is the full structure of each `job_workflow_ref`:

```
woodruffw/gha-oidc-jwt-test-repo/.github/workflows/release.yml@refs/tags/v0.0.3
```

The `username/repo` prefix and `.github/workflows` component are what makes the `startswith` check safe -- a malicious user can't contrive a workflow name such that those would pass, since workflows have to live inside a flat directory (`.github/workflows`).

I can definitely add some additional documentation here to explain the structure we're matching here!

Reply:
Hm, but then if matching the workflow name itself don't contribute to security, why do we require the name of the workflow at all ? I'm not asking that we change it all at this stage, but rather making sure I understand the security model.

Reply:
Sorry if I'm explaining poorly -- the workflow name *does* contribute to security. Here's the rough model:

*Identity*: The identity of the trusted repository is defined by `owner/repo` + `owner_id`, where `owner` is the GitHub owner identity (username/org name), `repo` is the repository name, and `owner_id` is the unique and never reused GitHub ID for `owner`.

*Authenticity*: The authenticity of a token is determined by its validity against GitHub's published JWK set. 

*Authority*: Only certain workflows are trusted for the purpose of minting access tokens from JWTs. We don't trust every workflow, because some workflows might be configured as runnable by untrusted users (e.g., on every commit). The workflow we trust for each provider setup is defined by `workflow.yml`, which is the unique name of the workflow specification in the repository.

Then, in terms of claim verification:

* Identity: `repository` ensures that we're validating the correct token.
* Authenticity: This is provided by the JWT payload verification iteself.
* Authority: `workflow` and `job_workflow_ref` ensure that we're only accepting tokens from our trusted workflow.

To summarize: `workflow` and `job_workflow_ref` are the two claims that ensure that we don't accept JWTs that are valid but aren't trusted for our purposes. They're technically slightly redundant against each other, but having both makes us a little more resilient against potential internal changes to GitHub (e.g., if GitHub one day allows workflows to override their name internally with a `workflow:` key, `job_workflow_ref` will still be safe because it contains the fully qualified repository path).

Reply:
I think what @ewjoachim is trying to say is that if the project has been configured to trust `user/repo/.github/workflows/foo.yml`, using `startswith` will also implicitly trust `user/repo/.github/workflows/foo.yml@malicious.yml`:

```
>>> expected_job_workflow_ref = "user/repo/.github/workflows/foo.yml@"
>>> claimed_job_workflow_ref = "user/repo/.github/workflows/foo.yml@malicious.yml"
>>> claimed_job_workflow_ref.startswith(expected_job_workflow_ref)
True
```

Reply:
(That 👆)

Reply:
Hm reading from https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/using-openid-connect-with-reusable-workflows it looks like the repository and the workflow ref might not match, so if we do it this way, it's possible that it won't be possible to use reusable workflows, no ?

Reply:
I believe this is correct, and that we should be able to support reusable workflows as well, so we should probably revisit this to allow a user to specify a workflow at a ref outside the repo.

Reply:
Sorry, was out for most of last week. I'll catch up here today!

Reply:
Re: `workflow` and `job_workflow` ref: I think we should always require *at least* one, since the absence of both makes it much easier for someone to trick a repository owner into merging changes to *any* workflow and getting direct access to the PyPI project.

That being said, I think you're both right about the two not necessarily matching. So yes, we should probably allow users to configure one or the other, but always at least one (and for `job_workflow_ref`, support globs). Does that seem reasonable?

Reply:
More re: `job_workflow_ref`: we should probably be careful about which positions we support globs in, since we don't want someone to accidentally configure `*/*` or even `foo/bar/.github/workflows/*.yml` (since that'll be the same insecure configuration as allowing any `workflow`).

Reply:
Let's not worry about supporting reusable workflows for now and assume the workflow lives in the repo (and file an issue capturing that deficiency).

I still think we need to resolve the use of `startswtih` here though.

Reply:
Sounds good. I'll mark `job_workflow_ref` as an explicitly unverified claim for now and document the plans there.

Reply:
Tracking here: https://github.com/pypa/warehouse/issues/11096

Reply:
04c72611